src/lib/loaders/__tests__/batchLoader.test.ts

```diff
@@ -0,0 +1,58 @@
+import { batchLoader } from "../batchLoader"
+
+describe("batchLoader", () => {
+  beforeEach(() => {
+    jest.resetModules()
+  })
+
+  describe("when batching is disabled", () => {
+    jest.doMock("config", () => ({
+      ENABLE_RESOLVER_BATCHING: "false",
+    }))
+
+    it("should return a single loader if one is provided", () => {
+      const singleLoader = () => {}
+      expect(batchLoader({ singleLoader, multipleLoader: null })).toBe(
+        singleLoader
+      )
+    })
+
+    it("should return a multipleLoader if no singleLoader provided", () => {
+      const multipleLoader = () => {}
+      expect(batchLoader({ multipleLoader })).toBe(multipleLoader)
+    })
+  })
+
+  it("should call the single loader when one item is batched", async () => {
+    const singleLoader = jest.fn(v => Promise.resolve(v))
+    const multipleLoader = jest.fn()
+    const batch = batchLoader({ singleLoader, multipleLoader })
+
+    await batch("foo")
+    expect(multipleLoader).not.toBeCalled()
+    expect(singleLoader).toBeCalledWith("foo")
+  })
+
+  /**
+   * A list endpoint (like sales) shouldn't provide a singleLoader because the
+   * multiple loader could support filtering options not supported by the single
+   * endpoint.
+   */
+  it("should call the multiple loader when one argument when no single loader provided", async () => {
+    const multipleLoader = jest.fn(v => Promise.resolve(v))
+    const batch = batchLoader({ multipleLoader })
+    const params = { id: "foo", param: "bar" }
+
+    await batch(params)
+    expect(multipleLoader).toBeCalledWith(params)
+  })
+
+  it("should group multiple calls together", async () => {
+    const multipleLoader = jest.fn(v => Promise.resolve(v))
+    const batch = batchLoader({ multipleLoader })
+
+    await Promise.all([batch("123"), batch("456"), batch("789")])
```

Comment:
DataLoader performs its work on the end of the runloop, so I bet you need to do something like:

```
expect.hasAssertions()
const batches = Promise.all([batch("123"), batch("456"), batch("789")])
setImmediate(() => batches.then(() => {
  expect(multipleLoader).toBeCalledTimes(1)
}))
```

Reply:
And otherwise take a look at the test-suite of DataLoader to see how they test them.

Reply:
Just saw your updated comment 😸